### PR TITLE
Optimize the log of the entropy coeff instead of the entropy coeff

### PR DESCRIPTION
**Pull request by jamesheald**
SB3 SAC optimizes the log of the entropy coeff instead of the entropy coeff ([link](https://github.com/DLR-RM/stable-baselines3/blob/512eea923afad6f6da4bb53d72b6ea4c6d856e59/stable_baselines3/sac/sac.py#L231)), as discussed [here](https://github.com/rail-berkeley/softlearning/issues/37). In contrast, SBX SAC optimizes the entropy coeff ([link](https://github.com/araffin/sbx/blob/19c85a1fd8fd56437c2f9c1de081889e49554604/sbx/sac/sac.py#L329)).

In my experience, this small change has a huge impact on the stability and performance of the algorithm (optimizing the log of the entropy coeff produces much better results than optimizing the entropy coeff).

### CODE REVIEW

**Comment by jan1854** [comment]
Hi, the SAC version in SBX already optimizes the log of the entropy coefficient. The `EntropyCoefficient` module contains a single parameter `log_ent_coef`, which is exponentiated when the module is called (see [here](https://github.com/araffin/sbx/blob/19c85a1fd8fd56437c2f9c1de081889e49554604/sbx/sac/sac.py#L27-L28)). Therefore, `ent_coef_value` [here](https://github.com/araffin/sbx/blob/19c85a1fd8fd56437c2f9c1de081889e49554604/sbx/sac/sac.py#L328) is the exponential of the log parameter (which is optimized).

**Comment by jan1854** [comment]
I see, the loss is indeed different from SB3. I think it could be a bit cleaner to change `EntropyCoef` to directly return the log entropy (basically removing the `jnp.exp()`) and exponentiating explicitly when updating the actor / critic. This would avoid accessing the parameters directly in `temperature_loss` (and it would be closer to the SB3 implementation). What do you think @araffin?

**Comment by araffin** [comment]
@jamesheald thanks for finding that out. It is a difference with SB3 indeed.

I have to say, I don't remember why I didn't directly store/returned the ent coef (probably the intention was to use the log in the loss but I forgot :see_no_evil: ).
I think a backward compatible solution would be to store and return the ent coef directly (instead of using `exp()` all the time) and convert old models `log_ent_coef` param automatically. What do you think @jan1854 ?

cc @danielpalen this should not but might affect CrossQ result slightly (I don't think too much because of https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/243 results).

@jamesheald do you have access to some compute to compare to current SBX master version? (I can grant you access to https://wandb.ai/openrlbenchmark/sbx if you want to push the runs too)

**Comment by jamesheald** [comment]
@araffin yes I have access to compute. Do you just want me to perform a few of runs of the modified SBX SAC on Walker2d-v4 using default parameters?

**Comment by araffin** [comment]
> Do you just want me to perform a few of runs of the modified SBX SAC on Walker2d-v4 using default parameters?

I would do on the classic 4 mujoco envs, see for instance https://wandb.ai/openrlbenchmark/sbx/reports/CrossQ-SBX-Perf-Report--Vmlldzo3MzQxOTAw

I'll also some runs on my machine (I only have access to one machine currently) and prepare a report.

**Comment by araffin** [comment]
Update: I started the report here: https://wandb.ai/openrlbenchmark/sbx/reports/SAC-SB3-vs-SBX-Ent-coef-vs-log-ent-coef--Vmlldzo5NjI3NTQ5 and will start the runs soon (I'm using the RL Zoo, see SBX readme)

My config looks like that:
```python
default_hyperparams = dict(
    n_envs=1,
    n_timesteps=int(1e6),
    policy="MlpPolicy",
    # qf_learning_rate=1e-3,
    policy_kwargs={},
    learning_starts=10_000,
)

hyperparams = {}

for env_id in [
    "HalfCheetah-v4",
    "Humanoid-v4",
    "HalfCheetahBulletEnv-v0",
    "Ant-v4",
    "Hopper-v4",
    "Walker2d-v4",
    "Swimmer-v4",
]:
    hyperparams[env_id] = default_hyperparams
```

and the command line (I'm modifying the train freq/gradient steps for faster execution):
```
python train.py --algo sac --env HalfCheetah-v4 -params n_envs:12 gradient_steps:12 -c sac_params.py -P \
--verbose 0 -f logs/ent_coef --vec-env subproc --n-eval-envs 5 --eval-episodes 20 --eval-freq 50000 \
-n 300000 --track --wandb-entity openrlbenchmark --wandb-project-name sbx -tags ent-coef v0.18.0
```

**Comment by jan1854** [comment]
> I think a backward compatible solution would be to store and return the ent coef directly (instead of using `exp()` all the time) and convert old models `log_ent_coef` param automatically. What do you think @jan1854 ?
> 
> EDIT: the easiest is just to take the log in the loss

For SAC, I think we would just need to call `exp()` [here](https://github.com/araffin/sbx/blob/19c85a1fd8fd56437c2f9c1de081889e49554604/sbx/sac/sac.py#L256) and [here](https://github.com/araffin/sbx/blob/19c85a1fd8fd56437c2f9c1de081889e49554604/sbx/sac/sac.py#L309), and this solution should be backwards compatible (as the name of the parameter does not change). We can of course also just call `log()` in the loss, but that would mean that we essentially do `log(exp(log_ent_coef))` in the entropy cofficient loss, which does not feel super clean.

**Comment by araffin** [comment]
> For SAC, I think we would just need to call exp() [here](https://github.com/araffin/sbx/blob/19c85a1fd8fd56437c2f9c1de081889e49554604/sbx/sac/sac.py#L256) and [here](https://github.com/araffin/sbx/blob/19c85a1fd8fd56437c2f9c1de081889e49554604/sbx/sac/sac.py#L309), and this solution should be backwards compatible (as the name of the parameter does not change).

I considered that but the issue was with both naming and with constant entropy coef (in that case, taking the exp is wrong).

> , but that would mean that we essentially do log(exp(log_ent_coef)) in the entropy cofficient loss, which does not f

I also thought the same but hope that jax jit may simplify this expression to identity?
see https://jax.readthedocs.io/en/latest/faq.html#jit-changes-the-exact-numerics-of-outputs
That also means changing things at one place only.

**Comment by jan1854** [comment]
True, we would probably need to rename `ent_coef_state` to `log_ent_coef_state` or something, which will cause problems when loading an old model. So your solution is probably the easiest.

**Comment by araffin** [comment]
I did many more runs, and looking at  https://wandb.ai/openrlbenchmark/sbx/reports/SAC-SB3-vs-SBX-Ent-coef-vs-log-ent-coef--Vmlldzo5NjI3NTQ5  I cannot see any significant different between SBX master (0.17.0) and this PR (0.18.0).

@jamesheald can you confirm that this PR solves your issue?
Can you provide a colab notebook or a minimal example (so minimal number of lines but everything needed to reproduce) to reproduce the original issue?

**Review by araffin — APPROVED** [review]
LGTM, thanks =)
